src/python/pants/notes/master.rst

```diff
@@ -4,6 +4,136 @@ Master Pre-Releases
 This document describes development releases which occur weekly from master, and which have
 not undergone the vetting associated with ``stable`` releases.
 
+1.16.0.dev2 (4/22/2019)
+-----------------------
+
+New features
+~~~~~~~~~~~~
+
+* Add support for v2-only goals, and replace list with a @console_rule (#6880)
+  `PR #6880 <https://github.com/pantsbuild/pants/pull/6880>`_
+
+* make deprecations respect warnings filters and add --ignore-pants-warnings option (#7496)
+  `PR #7496 <https://github.com/pantsbuild/pants/pull/7496>`_
+
+Bugfixes
+~~~~~~~~
+
+* Rerun `select-interpreter` if global Python interpreter constraints have changed  (#7586)
+  `PR #7586 <https://github.com/pantsbuild/pants/pull/7586>`_
+
+* Fix help info output for option choices (#7569)
+  `PR #7569 <https://github.com/pantsbuild/pants/pull/7569>`_
+
+* Directory materialization is visible to others. (#7562)
+  `PR #7562 <https://github.com/pantsbuild/pants/pull/7562>`_
+
+* Coerce argv for JavacCompile to text_type. (#7560)
+  `PR #7560 <https://github.com/pantsbuild/pants/pull/7560>`_
+
+* Fix doc generation rendering byte literals with Python 3 (#7556)
+  `PR #7556 <https://github.com/pantsbuild/pants/pull/7556>`_
+
+* fix go test flags bug and add integration test (#7539)
+  `PR #7539 <https://github.com/pantsbuild/pants/pull/7539>`_
+
+Documentation
+~~~~~~~~~~~~~
+
+* Fix link for common_task/target_aggregate (#7574)
+  `PR #7574 <https://github.com/pantsbuild/pants/pull/7574>`_
+
+* Update install guide to use `./pants generate-pants-ini` and `pants_runtime_python_version` (#7453)
+  `PR #7453 <https://github.com/pantsbuild/pants/pull/7453>`_
+
+Refactoring, Improvements, and Tooling
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* Hotfix #7591 breaking OSX Rust tests shard due to Pyenv global issue (#7602)
+  `PR #7602 <https://github.com/pantsbuild/pants/pull/7602>`_
+  `PR #7591 <https://github.com/pantsbuild/pants/pull/7591>`_
+
+* Pin OSX wheel release shards to pyenv. (#7591)
+  `PR #7591 <https://github.com/pantsbuild/pants/pull/7591>`_
+
+* Replace the `goals` goal with a `help` plugin which supports both v1 and v2 (#7598)
+  `PR #7598 <https://github.com/pantsbuild/pants/pull/7598>`_
+
+* Remove unused migrations folder (#7592)
+  `PR #7592 <https://github.com/pantsbuild/pants/pull/7592>`_
+
+* Collection subclasses correctly coerce to bools (#7584)
+  `PR #7584 <https://github.com/pantsbuild/pants/pull/7584>`_
+
+* Use ci.sh instead of travis-ci.sh (#7582)
+  `PR #7582 <https://github.com/pantsbuild/pants/pull/7582>`_
+
+* Cleanup run_python_test rule. (#7577)
+  `PR #7577 <https://github.com/pantsbuild/pants/pull/7577>`_
+
+* Improve `--pantsd-invalidation-globs` using Snapshot fingerprints (#7531)
+  `Issue #5567 <https://github.com/pantsbuild/pants/issues/5567>`_
+  `PR #7531 <https://github.com/pantsbuild/pants/pull/7531>`_
+
+* Restore support for pantsbuild.pants wheel using Python 3.7+ by removing abi3 workaround (#7578)
```

Comment:
Should this be a new feature?

Reply:
I think the PR title overstates the general case. We won't support python3.7, etc until we do one of these:
1. Have an audit process that proves pants.pex has 0 cp3X wheels (just abi3).
2. Have an audit process that proves pants.pex has all cp3X wheels where X is an element of {6, 7, ...} for any wheel that is platform specific but not abi3

It may be that right now we - by luck - support 3.6 and 3.7 because we depend on purely abi3 wheels where those are available. In the future we might depend on a wheel that does not use the limited abi and so can't be abi3.

Reply:
Good point. To guarantee Py37, we would need to add a Py37 wheel building shard, right? Which is currently blocked by Centos6 not supporting Py37, thus blocked by the upgrade to Centos7.

Reply:
I think that's all correct, yes.

Reply:
Yup, it looks like we're deep in the non-abi3 hole already.

Just one abi3 dep:
```
$ zipinfo ./pants.1.16.0.dev2.py36.pex | grep abi3 | cut -d/ -f2 | sort -u
cryptography-2.6.1-cp34-abi3-macosx_10_6_intel.whl
cryptography-2.6.1-cp34-abi3-manylinux1_x86_64.whl
pantsbuild.pants-1.16.0.dev2+git09c5c09e-cp36-abi3-linux_x86_64.whl
pantsbuild.pants-1.16.0.dev2+git09c5c09e-cp36-abi3-macosx_10_11_x86_64.whl
```

And many non:
```
$ zipinfo ./pants.1.16.0.dev2.py36.pex | grep cp36m | cut -d/ -f2 | sort -u
cffi-1.11.5-cp36-cp36m-macosx_10_6_intel.whl
cffi-1.11.5-cp36-cp36m-manylinux1_x86_64.whl
psutil-5.4.8-cp36-cp36m-linux_x86_64.whl
psutil-5.4.8-cp36-cp36m-macosx_10_11_x86_64.whl
pywatchman-1.4.1-cp36-cp36m-linux_x86_64.whl
pywatchman-1.4.1-cp36-cp36m-macosx_10_11_x86_64.whl
PyYAML-5.1-cp36-cp36m-linux_x86_64.whl
PyYAML-5.1-cp36-cp36m-macosx_10_11_x86_64.whl
scandir-1.2-cp36-cp36m-linux_x86_64.whl
scandir-1.2-cp36-cp36m-macosx_10_11_x86_64.whl
setproctitle-1.1.10-cp36-cp36m-linux_x86_64.whl
setproctitle-1.1.10-cp36-cp36m-macosx_10_11_x86_64.whl
thriftpy2-0.4.2-cp36-cp36m-linux_x86_64.whl
thriftpy2-0.4.2-cp36-cp36m-macosx_10_11_x86_64.whl
```

As such, we'll definitely need a wheel-builder shard and per-3.minor to support the pexes we burn. The venv based setups though should be able to run under 3.7+ with no intervention from us.